holoviews/plotting/bokeh/widgets.py

```diff
@@ -1,12 +1,208 @@
+from __future__ import unicode_literals
+
+import math
 import json
+from functools import partial
 
 import param
+import numpy as np
 from bokeh.io import _CommsHandle
 from bokeh.util.notebook import get_comms
+from bokeh.models.widgets import Select, Slider, AutocompleteInput, TextInput
+from bokeh.layouts import layout, gridplot, widgetbox, row, column
 
+from ...core import Store, NdMapping, OrderedDict
+from ...core.util import drop_streams, unique_array, isnumeric, wrap_tuple_streams
 from ..widgets import NdWidget, SelectionWidget, ScrubberWidget
 from .util import serialize_json
 
+
+
+class BokehServerWidgets(param.Parameterized):
+    """
+    BokehServerWidgets create bokeh widgets corresponding to all the
+    key dimensions found on a BokehPlot instance. It currently supports
+    to types of widgets sliders (which may be discrete or continuous)
+    and dropdown widgets letting you select non-numeric values.
+    """
+
+    basejs = param.String(default=None, precedence=-1, doc="""
+        Defines the local CSS file to be loaded for this widget.""")
+
+    extensionjs = param.String(default=None, precedence=-1, doc="""
+        Optional javascript extension file for a particular backend.""")
+
+    css = param.String(default=None, precedence=-1, doc="""
+        Defines the local CSS file to be loaded for this widget.""")
+
+    position = param.ObjectSelector(default='right',
+        objects=['right', 'left', 'above', 'below'])
+
+    sizing_mode = param.ObjectSelector(default='fixed',
+        objects=['fixed', 'stretch_both', 'scale_width',
+                 'scale_height', 'scale_both'])
+
+    width = param.Integer(default=200, doc="""
+        Width of the widget box in pixels""")
+
+    def __init__(self, plot, renderer=None, **params):
+        super(BokehServerWidgets, self).__init__(**params)
+        self.plot = plot
+        streams = []
+        for stream in plot.streams:
+            if any(k in plot.dimensions for k in stream.contents):
+                streams.append(stream)
+        self.dimensions, self.keys = drop_streams(streams,
+                                                  plot.dimensions,
+                                                  plot.keys)
+        if renderer is None:
+            backend = Store.current_backend
+            self.renderer = Store.renderers[backend]
+        else:
+            self.renderer = renderer
+        # Create mock NdMapping to hold the common dimensions and keys
+        self.mock_obj = NdMapping([(k, None) for k in self.keys],
+                                  kdims=self.dimensions)
+        self.widgets, self.lookups = self.get_widgets()
+        self.reverse_lookups = {d: {v: k for k, v in item.items()}
+                                for d, item in self.lookups.items()}
+        self.subplots = {}
+        if self.plot.renderer.mode == 'default':
+            self.attach_callbacks()
+        self.state = self.init_layout()
+        self._event_queue = []
```

Comment:
This is in widgets now so my original comment was addressed. Might be worth renaming here too - up to you.